### ISSUE TITLE
[BUG]

### ISSUE THREAD

**Issue by wzx226226** [flagos-ai/Torch-FL#252]
# [Performance] `aten::add.Tensor` via FlagGems is significantly slower than DAS boxing on DCU

## Description

On Hygon DCU, `aten::add.Tensor` produces correct results through both the Torch-FL DAS boxing route and the mixed FlagGems route. However, the FlagGems route shows a significant per-call performance regression for the small, non-contiguous, scalar, and broadcast inputs used by ESM-2.

The regression is reproducible with a standalone synchronized microbenchmark and is also visible during a complete ESM-2 embedding inference. Further investigation with `hipprof` indicates that the actual FlagGems DCU addition kernel is not slow. Most of the latency is spent outside the device kernel, in metadata processing, the Python/Triton wrapper, configuration lookup, argument preparation, and the host-side kernel launch path.

## Environment

- **Hardware**: Hygon DCU / BW1000
- **Microbenchmark node and device**: `a02r4n07`, physical device `0`; both boxing and FlagGems tests were run on this node and device
- **End-to-end ESM-2 evidence**: `a02r2n06`, physical device `0`; this was a separate same-device A/B experiment
- **Environment activation**: `source /work1/share/scnethpc2601/flagos/activate.sh`
- **Python**: 3.10
- **PyTorch**: `2.10.0+cpu`
- **HIP**: `6.3.26113`
- **Torch-FL configuration**: installed `backends_dcu_flaggems.conf`
- **Device visibility**: `HIP_VISIBLE_DEVICES=0`, `CUDA_VISIBLE_DEVICES=0`
- **Model workload**: ESM-2 8M embedding, three protein sequences, token shape `(3, 101)`, FP32

## Runtime Configuration

```bash
source /work1/share/scnethpc2601/flagos/activate.sh
export HIP_VISIBLE_DEVICES=0
export CUDA_VISIBLE_DEVICES=0
export FLAGOS_USE_FLAGGEMS=1
unset FLAGOS_BACKEND_CONFIG FLAGOS_USE_FLAGGEMS_CPP
```

The DAS boxing control was run in a separate Python process with `FLAGOS_USE_FLAGGEMS` unset. Both microbenchmark routes used the same node, physical device, input shapes, dtype, warmup count, repetition count, and synchronization method. The end-to-end ESM-2 results came from a separate same-device A/B experiment and should not be treated as the same measurement run as the standalone microbenchmark.

## Minimal Performance Reproduction

The following script contains all four test cases used in the comparison. They cover the non-contiguous residual input, non-contiguous attention input, scalar addition, and one-dimensional bias broadcasting observed in ESM-2. Each case is warmed up 30 times and then measured over 200 repetitions, with an explicit device synchronization after the measured loop.

Save it as `add_tensor_microbench.py`:

```python
import argparse
import time

import torch
import torch_fl


cases = {
    "residual_101x3x320_noncontig": (
        torch.randn(3, 101, 320, device="flagos:0").transpose(0, 1),
        torch.randn(101, 3, 320, device="flagos:0"),
    ),
    "attention_60x101x16_noncontig": (
        torch.randn(101, 60, 16, device="flagos:0").transpose(0, 1),
        torch.randn(60, 101, 16, device="flagos:0"),
    ),
    "ffn_101x3x1280_scalar": (
        torch.randn(101, 3, 1280, device="flagos:0"),
        1.0,
    ),
    "bias_3x101x33": (
        torch.randn(3, 101, 33, device="flagos:0"),
        torch.randn(33, device="flagos:0"),
    ),
}

parser = argparse.ArgumentParser()
parser.add_argument("--warmup", type=int, default=30)
parser.add_argument("--repeat", type=int, default=200)
args = parser.parse_args()


def run_case(left, right, warmup, repeat):
    for _ in range(warmup):
        torch.ops.aten.add.Tensor(left, right, alpha=1)
    torch.flagos.synchronize()

    started = time.perf_counter()
    for _ in range(repeat):
        output = torch.ops.aten.add.Tensor(left, right, alpha=1)
    torch.flagos.synchronize()

    elapsed_us = (time.perf_counter() - started) * 1e6 / repeat
    assert torch.isfinite(output).all().item()
    return elapsed_us


for name, (left, right) in cases.items():
    elapsed_us = run_case(left, right, args.warmup, args.repeat)
    print(f"{name}: {elapsed_us:.3f} us/call")
```

The same script must be run in two separate Python processes because the Torch-FL backend route is selected when `torch_fl` is imported. The backend should not be changed inside an already initialized process.

### DAS boxing control

Run the following command with `FLAGOS_USE_FLAGGEMS` unset so that `aten::add.Tensor` uses the DAS CUDA boxing route:

```bash
ssh a02r4n07
source /work1/share/scnethpc2601/flagos/activate.sh
export HIP_VISIBLE_DEVICES=0
export CUDA_VISIBLE_DEVICES=0
unset FLAGOS_USE_FLAGGEMS
unset FLAGOS_USE_FLAGGEMS_CPP
unset FLAGOS_BACKEND_CONFIG

python -u add_tensor_microbench.py --warmup 30 --repeat 200
```

### FlagGems test

After the boxing process exits, start a new process on the same node and physical device. Set `FLAGOS_USE_FLAGGEMS=1` so that `aten::add.Tensor` uses the FlagGems Python/Triton route:

```bash
ssh a02r4n07
source /work1/share/scnethpc2601/flagos/activate.sh
export HIP_VISIBLE_DEVICES=0
export CUDA_VISIBLE_DEVICES=0
export FLAGOS_USE_FLAGGEMS=1
unset FLAGOS_USE_FLAGGEMS_CPP
unset FLAGOS_BACKEND_CONFIG

python -u add_tensor_microbench.py --warmup 30 --repeat 200
```

The inputs, test cases, warmup count, repetition count, and synchronization method are identical between the two commands. The only route-affecting difference is whether `FLAGOS_USE_FLAGGEMS` is set.

## Microbenchmark Results

The table reports the synchronized batch-average latency per call after 30 warmup iterations and 200 measured iterations. All values are in microseconds.

| Test case | Input shape/layout | DAS boxing | FlagGems | Additional latency | FlagGems / boxing |
| --- | --- | ---: | ---: | ---: | ---: |
| Residual addition | `(101,3,320)`, first input non-contiguous | `15.497 µs` | `309.517 µs` | `+294.020 µs` | `19.97x` |
| Attention addition | `(60,101,16)`, first input non-contiguous | `15.481 µs` | `305.277 µs` | `+289.796 µs` | `19.72x` |
| FFN scalar addition | `(101,3,1280) + 1.0` | `18.132 µs` | `336.548 µs` | `+318.416 µs` | `18.56x` |
| Bias broadcast addition | `(3,101,33) + (33,)` | `15.449 µs` | `292.517 µs` | `+277.068 µs` | `18.93x` |

All four test cases produced finite outputs. The reported interval contains 200 operator calls followed by one device synchronization, divided by 200. This avoids creating and recording a separate event pair for every individual operator call. Both routes completed successfully with exit code `0`.

## End-to-End ESM-2 Evidence

During one complete ESM-2 embedding forward pass, `aten::add.Tensor` is executed 33 times. The following table shows the aggregated per-overload measurements from that forward pass:

| Route | Total latency for 33 calls | Average latency per call |
| --- | ---: | ---: |
| DAS boxing | `1680.9 µs` | `50.94 µs` |
| FlagGems | `11846.0 µs` | `358.97 µs` |
| Additional FlagGems latency | `10165.1 µs` | `308.03 µs` |
| FlagGems / boxing | `7.047x` | `7.047x` |

Overriding only `aten::add.Tensor` to use boxing reduced the ESM-2 embedding median end-to-end latency from `42.663 ms` to `33.520 ms`, an improvement of approximately `9.143 ms`. The FlagGems and boxing outputs passed the numerical comparison with `rtol=1e-4` and `atol=1e-5`, so the performance difference is not caused by divergent results.

## Internal Stage Investigation

For the representative non-contiguous residual input `(101,3,320)`, the following steady-state timings were collected after warmup from the FlagGems implementation:

| FlagGems stage | Latency | Work performed |
| --- | ---: | --- |
| `pointwise_dynamic.prepare_args`, corresponding to `op.meta()` | approximately `93 µs` | Tensor/scalar validation, dtype promotion, broadcast-shape calculation, non-contiguous stride analysis, output preparation, and `StridedBuffer` construction |
| `instantiate` | approximately `2.7 µs` | Lookup of an already generated overload; this rules out repeated Triton compilation as the main cause |
| Generated Triton overload, within `op.impl()` | approximately `166 µs` | Triton wrapper execution, configuration-key and cache lookup, grid calculation, argument packing, queue submission, and kernel launch |
| Actual `add_func_kernel_rank_3` | approximately `4.374 µs` | Device-side elementwise addition measured with `hipprof` |

Relevant source locations in the repository are:

- `FlagGems/src/flag_gems/utils/pointwise_dynamic.py`
- `FlagGems/src/flag_gems/utils/libentry.py`
- `FlagGems/src/flag_gems/ops/add.py`

The actual DCU kernel takes only approximately `4.374 µs`, while the complete FlagGems `aten::add.Tensor` call takes approximately `358.97 µs` in the ESM-2 operator profile. This indicates that the regression is dominated by fixed host-side work outside the arithmetic kernel rather than by insufficient device-side addition throughput.

## Expected Behavior

For small elementwise additions, the FlagGems route should provide steady-state latency comparable to DAS boxing, or the backend should provide a general mechanism that avoids paying the full Python/Triton dispatch cost for every invocation. In the current measurements, the actual kernel takes approximately `4.374 µs`, whereas the complete ATen call takes approximately `358.97 µs`, indicating excessive host-side overhead.

## Possible Backend Fix Directions

1. Cache `op.meta()` results for repeated dtype, shape, stride, and layout signatures, including broadcasted shapes, strides, and output metadata.
2. Cache complete launch metadata for generated overloads, including the configuration key, grid, and argument descriptors, instead of rebuilding it on every invocation.
3. Reduce repeated work in the `op.impl()` Python wrapper and the C++/Python bridge; consider a backend C++ fast path for common binary elementwise operators.
4. Add a replay benchmarker for the DCU backend, or skip unnecessary event-timing autotuning after the configuration for a signature is already known.

These are intended as Torch-FL/FlagGems backend changes. Reproducing or addressing this issue should not require users to modify model code.

## Regression Test Proposal

A DCU performance regression test could cover the four representative cases above and perform the following checks:

- Verify that both boxing and FlagGems outputs contain only finite values.
- Compare numerical results between the two routes with an explicit tolerance.
- Record synchronized per-call latency using identical warmup and repetition counts.
- Report or fail when the FlagGems/boxing ratio exceeds a project-defined threshold.

The performance threshold should remain configurable per hardware platform. On the tested DCU system, a ratio above `2x` could initially be treated as a regression signal rather than as a universal fixed threshold.

## Checklist

- [x] Compared DAS boxing and FlagGems on the same node and physical device
- [x] Included a self-contained minimal performance reproduction
- [x] Covered four input patterns observed in a real workload
- [x] Included end-to-end ESM-2 evidence
- [x] Included ATen-level, FlagGems internal-stage, and `hipprof` kernel evidence
- [x] Verified finite outputs and numerical consistency
- [x] Suggested backend-level fix directions and a regression test

**Comment by lvyufeng** [comment]
标题？

**Comment by wzx226226** [comment]
？怎么了






万彰雄
***@***.***
